api/agent.go

```diff
@@ -288,6 +290,119 @@ func (a *Agent) Monitor(stopCh <-chan struct{}, q *QueryOptions) (<-chan *Stream
 	return frames, errCh
 }
 
+// CPUProfile returns a runtime/pprof cpu profile for a given server or node.
+// The profile will run for the amount of seconds passed in or default to 1.
+// If no serverID or nodeID are provided the current Agents server will be
+// used.
+//
+// The parameters are:
+// * serverID: server ID or name to query, also accepts "leader"
+// * nodeID: client node ID to query
+// * seconds: the amount of time to run the trace for.
+//
+// The call blocks until the profile finishes, and returns the raw bytes of the
+// profile.
+func (a *Agent) CPUProfile(serverID, nodeID string, seconds int, q *QueryOptions) ([]byte, error) {
+	if q == nil {
+		q = &QueryOptions{}
+	}
+	if q.Params == nil {
+		q.Params = make(map[string]string)
+	}
+
+	q.Params["seconds"] = strconv.Itoa(seconds)
+	q.Params["node_id"] = nodeID
+	q.Params["server_id"] = serverID
+
+	body, err := a.client.rawQuery("/v1/agent/pprof/profile", q)
+	if err != nil {
+		return nil, err
+	}
+
+	var resp []byte
+	resp, err = ioutil.ReadAll(body)
+	if err != nil {
+		return nil, err
+	}
+
+	return resp, nil
+}
+
+// Trace returns a runtime/pprof trace for a given server or node.
+// The trace will run for the amount of seconds passed in or default to 1.
+// If no serverID or nodeID are provided the current Agents server will be
+// used.
+//
+// The parameters are:
+// * serverID: server ID or name to query, also accepts "leader"
+// * nodeID: client node ID to query
+// * seconds: the amount of time to run the trace for.
+//
+// The call blocks until the profile finishes, and returns the raw bytes of the
+// profile.
+func (a *Agent) Trace(serverID, nodeID string, seconds int, q *QueryOptions) ([]byte, error) {
+	if q == nil {
+		q = &QueryOptions{}
+	}
+	if q.Params == nil {
+		q.Params = make(map[string]string)
+	}
+
+	q.Params["seconds"] = strconv.Itoa(seconds)
+	q.Params["node_id"] = nodeID
+	q.Params["server_id"] = serverID
+
+	body, err := a.client.rawQuery("/v1/agent/pprof/trace", q)
+	if err != nil {
+		return nil, err
+	}
+
+	var resp []byte
+	resp, err = ioutil.ReadAll(body)
+	if err != nil {
+		return nil, err
+	}
+
+	return resp, nil
+
+}
+
+// Profile returns a runtime/pprof profile using pprof.Lookup to determine
+// which profile to run. Accepts a client or server ID but not both simultaneously.
+//
+// The parameters are:
+// * serverID: server ID or name to query, also accepts "leader"
+// * nodeID: client node ID to query
+// * profile: the name of the runtime/pprof profile to lookup and run.
+// * debug: flag to specify if the profile should return human readable output.
+//
+// The call blocks until the profile finishes, and returns the raw bytes of the
+// profile.
+func (a *Agent) Profile(serverID, nodeID, profile string, debug int, q *QueryOptions) ([]byte, error) {
```

Comment:
Maybe DRY here - the functions are very similar - but they are small so up to you.

Reply:
I was thinking from the caller perspective it may be nice to have multiple instead of a more complex request signature or request  struct

Reply:
unsure what the right thing is and it's nbd.  In my mind, we can dry on implementation side with using a helper.

Not sure what the user prospective, we should note that `Profile` is the generic function and can handle any profile type whether cpu/memory or a "second-class" profile (e.g. heap/mutex).

Also, it might make sense to have `PProfOptions` - seconds here would be nice too.

Reply:
`LookupProfile` may be a better name for Profile, I'll dry up the implementations and add PprofOptions